File: src/gmeow_tools/diagnostics_config.py
# SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
# SPDX-License-Identifier: AGPL-3.0-only

"""Shared diagnostics output policy (#662).

One :class:`DiagnosticsConfig` owns *where* diagnostics go and *how* they are
projected to the console — console mode, which artifact files to write, the
output directory, filename stem, and the stable code-scanning category. It is
the single rail `gmeow-dev`, Make, and GitHub Actions all set policy through:
the same five knobs as CLI flags AND as ``GMEOW_DIAGNOSTICS_*`` environment
variables, resolved with one precedence — **flag > env > default**.

This module is pure policy: it touches no Rust, renders nothing, and never
changes a gate's exit code. ``silent`` and ``none`` suppress *output*; they never
swallow an error. Invalid tokens are a hard failure (no silent fallback), per the
no-optionality doctrine — ``auto``/``silent``/``none`` are explicitly enumerated,
validated modes, not degraded fallbacks.
"""

from __future__ import annotations

import os
import sys
from collections.abc import Mapping
from dataclasses import dataclass
from enum import StrEnum
from pathlib import Path

from gmeow_tools.config import DIST_DIR

#: The three artifact projections, in deterministic write order. Mirrors the
#: fixed order the Rust ``Report.write_artifacts`` honors regardless of selection.
ARTIFACT_KINDS: tuple[str, ...] = ("json", "sarif", "html")

DEFAULT_STEM = "gmeow-feedback"
DEFAULT_CATEGORY = "gmeow"


class ConsoleMode(StrEnum):
    """How findings are projected to the console.

    ``auto`` is resolved away during :meth:`DiagnosticsConfig.resolve` (to
    ``pretty`` on a TTY, ``text`` otherwise), so a resolved config never carries
    ``auto`` — the renderer only ever sees a concrete mode.
    """

    AUTO = "auto"
    PRETTY = "pretty"
    TEXT = "text"
    JSONL = "jsonl"
    SILENT = "silent"


def _first(*candidates: str | None) -> str:
    """Return the first non-``None`` candidate — the flag > env > default rule.

    The final candidate is always the default, so the result is never ``None``.
    """
    for candidate in candidates:
        if candidate is not None:
            return candidate
    raise AssertionError("at least one candidate (the default) must be non-None")


def _parse_artifacts(raw: str) -> frozenset[str]:
    """Parse an ``--diagnostics-artifacts`` value into a set of kinds.

    ``none`` -> empty, ``all`` -> every kind, otherwise a comma list of kinds.
    An unknown token is a hard error (a typo must not silently degrade to a
    narrower or empty selection).
    """
    token = raw.strip().lower()
    if token == "none":
        return frozenset()
    if token == "all":
        return frozenset(ARTIFACT_KINDS)
    kinds = {part.strip().lower() for part in token.split(",") if part.strip()}
    unknown = kinds - set(ARTIFACT_KINDS)
    if unknown:
        raise ValueError(
            f"unknown diagnostics artifact kind(s): {sorted(unknown)} "
            f"(expected a subset of {list(ARTIFACT_KINDS)}, or 'none'/'all')"
        )
    if not kinds:
        raise ValueError(f"empty diagnostics artifact selection: {raw!r}")
    return frozenset(kinds)


def _env_path(value: str | None) -> Path | None:
    """An env-supplied directory as a ``Path``, or ``None`` when unset/blank."""
    if value is None or not value.strip():
        return None
    return Path(value).expanduser()


@dataclass(frozen=True, slots=True)
class DiagnosticsConfig:
    """Resolved diagnostics output policy (immutable)."""

    console: ConsoleMode
    artifacts: frozenset[str]
    directory: Path
    stem: str
    category: str

    @classmethod
    def resolve(
        cls,
        *,
        console: str | None = None,
        artifacts: str | None = None,
        directory: Path | None = None,
        stem: str | None = None,
        category: str | None = None,
        env: Mapping[str, str] | None = None,
        is_tty: bool | None = None,
    ) -> DiagnosticsConfig:
        """Resolve the output policy from flags, environment, and defaults.

        Precedence is **flag > env > default** for every knob. ``auto`` resolves
        by ``is_tty`` (defaulting to ``sys.stderr.isatty()`` — the diagnostic
        console is the stderr surface, so piping stdout does not flip it). Invalid
        ``console``/``artifacts`` tokens raise rather than fall back.
        """
        env = os.environ if env is None else env
        if is_tty is None:
            is_tty = sys.stderr.isatty()

        mode = ConsoleMode(
            _first(console, env.get("GMEOW_DIAGNOSTICS_CONSOLE"), ConsoleMode.AUTO)
        )
        if mode is ConsoleMode.AUTO:
            mode = ConsoleMode.PRETTY if is_tty else ConsoleMode.TEXT

        resolved_artifacts = _parse_artifacts(
            _first(artifacts, env.get("GMEOW_DIAGNOSTICS_ARTIFACTS"), "all")
        )

        resolved_category = _first(
            category, env.get("GMEOW_DIAGNOSTICS_CATEGORY"), DEFAULT_CATEGORY
        )
        resolved_stem = _first(stem, env.get("GMEOW_DIAGNOSTICS_STEM"), DEFAULT_STEM)

        # Directory precedence: an explicit flag or env dir is used verbatim.
        # Otherwise the default is keyed on whether a *category* was explicitly
        # requested: an aggregate/manual run (no category) keeps the flat
        # ``dist/`` convention (preserving ``dist/gmeow-feedback.*``), while a
        # category run (CI per-job, ``--diagnostics-category lint``) lands under
        # ``dist/diagnostics/<category>/`` so per-job artifacts never collide.
        explicit_dir = directory or _env_path(env.get("GMEOW_DIAGNOSTICS_DIR"))
        category_explicit = category is not None or bool(
            env.get("GMEOW_DIAGNOSTICS_CATEGORY")
        )
        if explicit_dir is not None:
            resolved_dir = explicit_dir
        elif category_explicit:
            resolved_dir = DIST_DIR / "diagnostics" / resolved_category
        else:
            resolved_dir = DIST_DIR

        return cls(
            console=mode,
            artifacts=resolved_artifacts,
            directory=resolved_dir,
            stem=resolved_stem,
            category=resolved_category,
        )

File: crates/cli-core/tests/diagnostics_config.rs
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: AGPL-3.0-only

//! Rust twin of the retired `tests/test_diagnostics_config.py`.
//!
//! Covers the same 15 policy cases for the resolved diagnostics output config.

use std::collections::{BTreeSet, HashMap};
use std::path::PathBuf;

use gmeow_cli_core::error::{UnknownArtifactKind, UnknownConsoleMode};
use gmeow_cli_core::{ConsoleMode, DiagnosticsConfig};

fn dist_dir() -> PathBuf {
    PathBuf::from("dist")
}

fn env_empty() -> HashMap<String, String> {
    HashMap::new()
}

#[test]
fn defaults_resolve_with_no_flags_or_env() {
    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        None,
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.console, ConsoleMode::Pretty);
    assert_eq!(config.artifacts, btreeset(&["json", "sarif", "html"]));
    assert_eq!(config.directory, dist_dir());
    assert_eq!(config.stem, "gmeow-feedback");
    assert_eq!(config.category, "gmeow");
}

#[test]
fn auto_console_resolves_by_tty() {
    let on_tty = DiagnosticsConfig::resolve(
        Some("auto"),
        None,
        None,
        None,
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(on_tty.console, ConsoleMode::Pretty);

    let off_tty = DiagnosticsConfig::resolve(
        Some("auto"),
        None,
        None,
        None,
        None,
        &env_empty(),
        false,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(off_tty.console, ConsoleMode::Text);
}

#[test]
fn flag_beats_env_for_console() {
    let mut env = env_empty();
    env.insert("GMEOW_DIAGNOSTICS_CONSOLE".to_owned(), "silent".to_owned());
    let config = DiagnosticsConfig::resolve(
        Some("pretty"),
        None,
        None,
        None,
        None,
        &env,
        false,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.console, ConsoleMode::Pretty);
}

#[test]
fn env_honored_when_no_flag() {
    let mut env = env_empty();
    env.insert("GMEOW_DIAGNOSTICS_CONSOLE".to_owned(), "jsonl".to_owned());
    let config =
        DiagnosticsConfig::resolve(None, None, None, None, None, &env, true, &dist_dir()).unwrap();
    assert_eq!(config.console, ConsoleMode::Jsonl);
}

#[test]
fn stem_precedence() {
    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        Some("foo"),
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.stem, "foo");

    let mut env = env_empty();
    env.insert("GMEOW_DIAGNOSTICS_STEM".to_owned(), "bar".to_owned());
    let config =
        DiagnosticsConfig::resolve(None, None, None, None, None, &env, true, &dist_dir()).unwrap();
    assert_eq!(config.stem, "bar");

    let config =
        DiagnosticsConfig::resolve(None, None, None, Some("foo"), None, &env, true, &dist_dir())
            .unwrap();
    assert_eq!(config.stem, "foo");

    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        None,
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.stem, "gmeow-feedback");
}

#[test]
fn category_precedence() {
    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        None,
        Some("lint"),
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.category, "lint");

    let mut env = env_empty();
    env.insert("GMEOW_DIAGNOSTICS_CATEGORY".to_owned(), "rust".to_owned());
    let config =
        DiagnosticsConfig::resolve(None, None, None, None, None, &env, true, &dist_dir()).unwrap();
    assert_eq!(config.category, "rust");

    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        None,
        Some("lint"),
        &env,
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.category, "lint");

    let config = DiagnosticsConfig::resolve(
        None,
        None,
        None,
        None,
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .unwrap();
    assert_eq!(config.category, "gmeow");
}

#[test]
fn artifacts_parsing() {
    let cases: &[(&str, &[&str])] = &[
        ("all", &["json", "sarif", "html"]),
        ("none", &[]),
        ("json,sarif", &["json", "sarif"]),
        ("sarif,json", &["json", "sarif"]),
        ("HTML", &["html"]),
    ];
    for (raw, expected) in cases {
        let config = DiagnosticsConfig::resolve(
            None,
            Some(raw),
            None,
            None,
            None,
            &env_empty(),
            true,
            &dist_dir(),
        )
        .unwrap();
        assert_eq!(config.artifacts, btreeset(expected), "artifacts={raw}");
    }
}

#[test]
fn unknown_artifact_token_hard_fails() {
    for raw in ["json,xml", "pdf", "json,,sarif,bogus"] {
        let err = DiagnosticsConfig::resolve(
            None,
            Some(raw),
            None,
            None,
            None,
            &env_empty(),
            true,
            &dist_dir(),
        )
        .expect_err(&format!("raw={raw} must hard-fail"));
        // The hard failure is the substrate's own downcastable kind, carrying the
        // stable registered code — not a bespoke error enum.
        assert!(
            err.is::<UnknownArtifactKind>(),
            "raw={raw}: expected UnknownArtifactKind, got {err}"
        );
        assert_eq!(err.code(), UnknownArtifactKind::register(), "raw={raw}");
    }
}

#[test]
fn invalid_console_token_hard_fails() {
    let err = DiagnosticsConfig::resolve(
        Some("loud"),
        None,
        None,
        None,
        None,
        &env_empty(),
        true,
        &dist_dir(),
    )
    .expect_err("an unknown console token must hard-fail");
    assert!(
        err.is::<UnknownConsoleMode>(),
        "expected UnknownConsoleMode, got {err}"
    );
    assert_eq!(err.code(), UnknownConsoleMode::register());
}

#[test]
fn directory_default_is_flat_dist_without_a_category() {
    for is_tty in [true, false] {
        let config = DiagnosticsConfig::resolve(
            None,
            None,
            None,
            None,
            None,
            &env_empty(),
            is_tty,
            &dist_dir(),
        )
        .unwrap();
        assert_eq!(config.directory, dist_dir());
    }
}

#[test]
fn directory_is_category_scoped_when_category_explicit() {
    for is_tty in [true, false] {
        let config = DiagnosticsConfig::resolve(
            None,
            None,
            None,
            None,
            Some("lint"),
            &env_empty(),
            is_tty,
            &dist_dir(),
        )
        .unwrap();
        assert_eq!(
            config.directory,
            dist_dir().join("diagnostics").join("lint")
        );
    }
}

#[test]
fn directory_is_category_scoped_when_category_from_env() {
    let mut env = env_empty();
    env.insert("GMEOW_DIAGNOSTICS_CATEGORY".to_owned(), "rust".to_owned());
    let config =
        DiagnosticsConfig::resolve(None, None, None, None, None, &env, true, &dist_dir()).unwrap();
    assert_eq!(
        config.directory,
        dist_dir().join("diagnostics").join("rust")
    );
}

#[test]
fn explicit_directory_flag_wins_in_both_modes() {
    let explicit = PathBuf::from("/tmp/explicit-dir");
    for is_tty in [true, false] {
        let config = DiagnosticsConfig::resolve(
            None,
            None,
            Some(&explicit),
            None,
            Some("lint"),
            &env_empty(),
            is_tty,
            &dist_dir(),
        )
        .unwrap();
        assert_eq!(config.directory, explicit);
    }
}

#[test]
fn env_directory_wins_over_default_off_tty() {
    let env_dir = PathBuf::from("/tmp/env-dir");
    let mut env = env_empty();
    env.insert(
        "GMEOW_DIAGNOSTICS_DIR".to_owned(),
        env_dir.to_str().unwrap().to_owned(),
    );
    let config =
        DiagnosticsConfig::resolve(None, None, None, None, None, &env, false, &dist_dir()).unwrap();
    assert_eq!(config.directory, env_dir);
}

#[test]
fn is_tty_falls_back_to_stderr_terminal() {
    let is_tty = std::io::IsTerminal::is_terminal(&std::io::stderr());
    let config = DiagnosticsConfig::resolve(
        Some("auto"),
        None,
        None,
        None,
        None,
        &env_empty(),
        is_tty,
        &dist_dir(),
    )
    .unwrap();
    let expected = if is_tty {
        ConsoleMode::Pretty
    } else {
        ConsoleMode::Text
    };
    assert_eq!(config.console, expected);
}

fn btreeset(items: &[&str]) -> BTreeSet<String> {
    items.iter().map(|s| s.to_string()).collect()
}
